### ISSUE TITLE
Hercules REXX initialization fails at Hercules startup for ooRexx version 5

### ISSUE THREAD

**Issue by ghost** [SDL-Hercules-390/hyperion#301]
ooRexx version 5 does not provide any longer the librexxutil dynamic library
the new library  librxunixsys which provides the same facilities is already loaded by rexx

**Comment by Fish-Git** [comment]
* **[How to Ask for Help](https://github.com/Fish-Git/AskForHelp)**

**Comment by 00003481** [comment]
I was not asking for any help 
I was only trying to make the hercules developers aware of the changes in ooRexx 

the librexxutil sources were  deleted from the repository 

r11758 | bigrixx | 2019-02-15 00:16:28 +0100 (Fri, 15 Feb 2019) | 1 line
Changed paths:
...
   D /main/trunk/extensions/rexxutil/platform/unix/rexxutil.cpp
   D /main/trunk/extensions/rexxutil/platform/windows/rexxutil.cpp
...

the build of the library was stopped quite a bit earlier

( the ooRexx developers are a bit conservative about deleting things from the repository )

librxunixsys provides the same facilities and is automatically loaded by the rexx initialization
so  the load the librexxutil.<dll/so/dylib> for ooRexx 5.0.0  at the current version will just yield a startup error 

[
...
HHC17531E REXX(OORexx) dlopen 'librexxutil.dylib' failed: dlopen(librexxutil.dylib, 9): image not found
HHC00007I Previous message from function 'OORexxLoadExtra' at hRexxapi.c(1273)
HHC17511E REXX() Could not enable either Rexx package
HHC00007I Previous message from function 'Enable' at hRexx.c(514)
HHC17511E REXX() Could not enable default Rexx package
HHC00007I Previous message from function 'InitializeRexx' at hRexx.c(261)             ...
]

all the information requested in the how to is IRRELEVANT to the current issue

the quick and dirty fix ( the *ix section )

[
diff --git a/hRexx_o.c b/hRexx_o.c
index d6c75875..54545dfc 100644
--- a/hRexx_o.c
+++ b/hRexx_o.c
@@ -82,10 +82,6 @@ static char*  REXX_DEP( ExtraLibs )[]  =
     "rxregexp.dll"          ,
     "rxsock.dll"            ,
     "rxwinsys.dll"          ,
-  #elif defined( __APPLE__ )
-    "librexxutil.dylib"     ,
-  #else
-    "librexxutil.so"        ,
   #endif
 };
]

the  warning
[
In file included from ../Hyperion.SDL/hRexx_o.c:93:
../Hyperion.SDL/hRexxapi.c:1259:43: warning: excess elements in array
      initializer
    void*  libhandle[ NUM_EXTRALIBS ]  = {0};
                                          ^
]
can be probably ignored

and hercules comes up with a working ooRexx interface 

systems  with a running ooRexx 
TESTED and FIXED
Apple/Darwin family - El Capitan, High Sierra, Mojave
Linux Family - Fedora 32, Centos 8.1,  Ubuntu 20
IN PROGRESS - just to spend time in the covid time-space continuum
BSD Family - FreeBSD 13, NetBSD 8, OpenBSD 6 , Dragonfly 5

**Comment by Fish-Git** [comment]
_Thank you_, Enrico! Your detailed explanation of the problem is very helpful.

I'm not sure why you closed this issue however, because the problem has certainly _not_ been resolved yet! So I'm going to re-open this issue until we can get it fixed.

Your suggested quick-fix at least tells us _where_ the problem is, so thank you for that! Much appreciated!

I'm a little stuck on what the _best_ way would be for fixing this problem however. I mean, _ultimately_ we need to somehow know what version of ooRexx the user has installed, so that we can bypass trying to load the `ExtraLibs` for ooRexx v5 or greater, but to continue trying to load them as we are today for ooRexx v4 and earlier.

Do you know if there is any type of "version" constant in any of the ooRexx headers that we can maybe check?

**Comment by Fish-Git** [comment]
_OR..._ (thinking out loud here), maybe all I need to do is treat the failure to load the `ExtraLibs` as a **non-fatal** "warning" type failure?

That is to say, break out the loading of the "Primary" (required) libraries and the loading  of the "Extra" libraries (non-required) into two separate functions (right now they're both being done by the `REXX_DEP( Load )` function, called by the `REXX_DEP( Enable )` function), such that if the loading of the Primary _required_ libraries fails, then Rexx support fails, but if the loading of the Extra _non-required_ libraries fails, it only causes warning messages to be issued but the `REXX_DEP( Enable )` function still returns success (instead of failing like it does today)?

Hmmm... I'll have to try that...

**Comment by 00003481** [comment]
now that I know a bit more about the ooRexx internals I  think hat loading the extra libraries is pretty useless  ,

librexx and librexxapi are generic dynamic libraries and the address resolution is done at load time

the external functions ( from the ooRexx POV ) libraries are a different can of worms  as far as the address resolution is concerned the script MUST use esplicit  functions / directives
  
there are two types of external libraries classic and native

for the classic external functions libraries written in C or C++
the dynamic library must contain user written functions to register and deregister  the  the additional functions  
Call RxFuncAdd "SysLoadFuncs", "RexxUTIL", "SysLoadFuncs"
Call SysLoadFuncs              /* call to load other functions    */

RXFUNCADD registers the external function SysLoadFuncs as routine SysLoadFuncs in the rexxutil library. SysLoadFuncs registers additional functions in rexxutil with RexxRegisterFunctionDll. See the SysLoadFuncs routine below for a function registration example.

so from a rexxutil point of view nothing changes as far as  the script code

for the native external function libraries C++ only 
( sequence of definitions coded according to the native api rules in the code )

the rexx script MUST contain the ::requires directive 
and registering of the additional function is in charge of the directive processor 

at the end of the long rant I hope to have clarified why the EXTAR loading is not needed

cheers
e

**Comment by Fish-Git** [comment]
> At the end of the long rant I hope to have clarified why the ExtraLibs loading is not needed.

The loading of the ExtraLibs is unfortunately very much needed. For details regarding the reason why, please refer to:

* [GitHub Issue #140 ("Random Hercules crash using OORexx with HAO")](https://github.com/SDL-Hercules-390/hyperion/issues/140)
* Specifically, [my Sep 6, 2018 comment](https://github.com/SDL-Hercules-390/hyperion/issues/140#issuecomment-419284634). It explains the ooRexx bug.
* And commit bd2a5ab03680afd16d9c3aa3ef8ba4536e49dc8f which introduced ExtraLibs in order to fix issue 140.

In short, ooRexx has a bug which can crash Hercules if the ExtraLibs are not pre-loaded. By pre-loading the ExtraLibs during Hercules Rexx support initialization, we prevent the ooRexx bug from occurring.

So we _are_ going to need to continue loading the ExtraLibs.

However, since the list of extra libraries currently defined in our `ExtraLibs` list clearly does not apply to ooRexx v5 (only to ooRexx v4 or earlier), we are unfortunately stuck between a rock and a hard place. &nbsp; `:(`

So as I said, we need to _somehow_ determine at runtime which version of ooRexx the user has installed. If they have v4 installed, then we proceed normally as we do today. If they have v5 installed however, then we need to skip the loading of the ExtraLibs (or perhaps load a _different_ set of libraries?)

What is the easiest way to determine _(**at runtime**)_ what version of ooRexx the user has installed?

_**OR...**_ As I suggested previously, I believe treating a failure to load ExtraLibs as only a _"warning"_ informational type error (and not a _fatal_ error like today) might also be a possible solution _(perhaps even the better one)._

What do you think?

**Comment by 00003481** [comment]
Hi Fish

I agree, Right now it might be the best approach.

But just out of curiosity , where can I find the original script that caused the error
I guess that was in a window environment , currently the hostemu dll is loaded only for windows


Technically the bug is not within Hercules -- it is within OORexx.
Not really IMO 
it just depends on how the rexx instance is created RexxStart in this case 
And that is the only option available for C 

And I discovered another limitation …
The argument parsing in the C style ( using the .SYSCARGS directory )
Is not available when using  RexxStart   

Using the native interface which unfortunately seems available only for C++ might solve everything

I will  run a few tests to find out  and let You know!

cheers
E

**Comment by 00003481** [comment]
Or… probably better
Since as part of the rexx initialization a rexxstart with “parse Version v;” Is issued You might process the output  
REXX-ooRexx_5.0.0(MT)_64-bits 6.05 06 May 2020 
To take the appropriate actions according to the rexx version

The above approach would work also for regina 

Since the loading of the external libraries is a preemptive action against the system dynamic loader 
And no rexx related process is done  it should be OK to issue the load even after the first successful rexx invocation

Cheers
E

PS a more object oriented approach would be to use the .rexxinfo class 

rexx -e "say .rexxinfo~version"       
5.0.0

**Comment by Fish-Git** [comment]
Enrico Sorichetti wrote:
> Fish wrote:
>
> > _**OR...**_ As I suggested previously, I believe treating a failure to load ExtraLibs as only a _"warning"_ informational type error (and not a _fatal_ error like today) might also be a possible solution _(perhaps even the better one?)._
> > 
> > What do you think?
> 
> I agree, Right now it might be the best approach.

Thanks.

Just to be safe I think maybe I'll do both. &nbsp; `;-)`
&nbsp;

> P.S. a more object oriented approach would be to use the `.rexxinfo` class:
> ```
>     rexx -e "say .rexxinfo~version"
>     5.0.0
> ```

I like that! &nbsp; `:)`
&nbsp;

> But just out of curiosity , where can I find the original script that caused the error?

The original script is attached to the original GitHub Issue problem report. It's called "**[Mount#scr.rex.txt](https://github.com/SDL-Hercules-390/hyperion/files/2335281/Mount.scr.rex.txt)**". Just click on that link to download it.

The script has something to do with his attempt to automate the mounting of scratch tapes whenever a mount message is seen. The script grabs the first entry in his list of scratch tapes (Lst-Scratch.txt) and uses that in a Hercules `devinit` command to automatically mount that tape.

After the script removes the first entry from the scratch list, it needs to then write that updated Windows file (Lst-Scratch.txt) back out to disk again. It uses Rexx's `EXECIO * DISKW` command to do that, and that is where the crash, whenever it would occur, would always occur. That is to say, the crash did not always occur. In fact it rarely occurred. But sometimes it would, and whenever it did occur, it was _always_ on the EXECIO statement.

> I guess that was in a Window environment, since currently, the `hostemu.dll` is loaded only for Windows.

Correct. The script contains a `::requires hostemu LIBRARY` directive at the end to ensure the `hostemu.dll` is loaded since that is where ooRexx's "EXECIO" logic is for Windows.

**Comment by 00003481** [comment]
Another Follow on
>     rexx -e "say .rexxinfo~version"
>     5.0.0

rexx -e "say .rexxinfo~majorversion"                    
5


Is even better :-)
> But just out of curiosity , where can I find the original script that caused the error?
> 

Already found it 


E

**Comment by Fish-Git** [comment]
> > > P.S. a more object oriented approach would be to use the `.rexxinfo` class:
> > > <pre>
> > >     rexx -e "say .rexxinfo~version"
> > >     5.0.0
> > > </pre>
> > 
> > I like that!   `:)`
>
> <pre>
>    rexx -e "say .rexxinfo~<i><b>majorversion</b></i>"
>    5
> </pre>
> Is even better. &nbsp; `:-)`

_**THANKS!**_

**Comment by Fish-Git** [comment]
Enrico (@00003481),

Please try the following patch and let me know whether or not it resolves your problem:

* [#301.patch.txt](https://github.com/SDL-Hercules-390/hyperion/files/4697436/301.patch.txt)

(here's what it looks like):

```diff
From: "User Name" <user@example.com>
Date: Thu, 28 May 2020 11:04:31 -0800
Subject: [PATCH] Short description

Details of change, perhaps spanning...
...multiple...
...lines.

diff -r -a -x .git -x 'msvc.AMD64.*' -x 'msvc.dllmod.*' -x 'msvc.debug.*' -x '*.suo' -x '*.ncb' -x '*.user' -x '*.htm' -x WORK -x DICTS -x FILES -x 'allTests.*' -x '*.rej' -x '*.orig' -x AutoBuildCount.h -x '*.cmp*' -x '*.comp*' -Nu hyperion-1/hRexx.c hyperion-0/hRexx.c
--- hyperion-1/hRexx.c	2020-05-18 12:26:39.263671800 -0700
+++ hyperion-0/hRexx.c	2020-05-28 10:59:53.970647100 -0700
@@ -59,6 +59,7 @@
 char*           PackageName       = NULL;
 char*           PackageVersion    = NULL;
 char*           PackageSource     = NULL;
+char            PackageMajorVers  = '0';
 BYTE            MsgLevel          = DEFAULT_MSGLEVEL;
 BYTE            MsgPrefix         = DEFAULT_MSGPREFIX;
 BYTE            ErrPrefix         = DEFAULT_ERRPREFIX;
diff -r -a -x .git -x 'msvc.AMD64.*' -x 'msvc.dllmod.*' -x 'msvc.debug.*' -x '*.suo' -x '*.ncb' -x '*.user' -x '*.htm' -x WORK -x DICTS -x FILES -x 'allTests.*' -x '*.rej' -x '*.orig' -x AutoBuildCount.h -x '*.cmp*' -x '*.comp*' -Nu hyperion-1/hRexx.h hyperion-0/hRexx.h
--- hyperion-1/hRexx.h	2020-05-18 12:26:39.264671900 -0700
+++ hyperion-0/hRexx.h	2020-05-28 10:54:46.535390400 -0700
@@ -62,6 +62,7 @@
 extern char*                PackageName;
 extern char*                PackageVersion;
 extern char*                PackageSource;
+extern char                 PackageMajorVers;
 extern BYTE                 MsgLevel;
 extern BYTE                 MsgPrefix;
 extern BYTE                 ErrPrefix;
diff -r -a -x .git -x 'msvc.AMD64.*' -x 'msvc.dllmod.*' -x 'msvc.debug.*' -x '*.suo' -x '*.ncb' -x '*.user' -x '*.htm' -x WORK -x DICTS -x FILES -x 'allTests.*' -x '*.rej' -x '*.orig' -x AutoBuildCount.h -x '*.cmp*' -x '*.comp*' -Nu hyperion-1/hRexxapi.c hyperion-0/hRexxapi.c
--- hyperion-1/hRexxapi.c	2020-05-18 12:26:39.266672000 -0700
+++ hyperion-0/hRexxapi.c	2020-05-28 10:59:25.091546400 -0700
@@ -1225,6 +1225,15 @@
         *source++ = 0; // (null terminate version and point to source)
         REXX_DEP( PackageVersion ) = strdup( Result );
         REXX_DEP( PackageSource  ) = strdup( source );
+
+        /* Extract the major version number too */
+
+        // "REXX-ooRexx_4.2.0(MT)_32-bit"
+        //              ^
+        //              ^
+        //              ^
+
+        REXX_DEP( PackageMajorVers ) = REXX_DEP( PackageVersion )[12];
     }
 
     return source ? TRUE : FALSE;
@@ -1269,7 +1278,7 @@
             if (verbose)
             {
                 // "REXX(%s) dlopen '%s' failed: %s"
-                WRMSG( HHC17531, "E", REXX_DEP( PackageName ),
+                WRMSG( HHC17531, "W", REXX_DEP( PackageName ),
                     libname, dlerror());
             }
         }
@@ -1388,12 +1397,19 @@
     /*---------------------------------------------------------------*/
     if (!REXX_DEP( LoadExtra )( verbose ))
     {
-        dlclose( LibHandle );
-        if (ApiLibHandle != LibHandle)
-            dlclose( ApiLibHandle );
-        LibHandle    = NULL;
-        ApiLibHandle = NULL;
-        return FALSE;
+        // A failure to load the ExtraLibs is fatal for
+        // version 4.2.0 of ooRexx or earlier. For ooRexx
+        // version 5.0.0 and later it is only a warning.
+
+        if (REXX_DEP( PackageMajorVers ) < '5')
+        {
+            dlclose( LibHandle );
+            if (ApiLibHandle != LibHandle)
+                dlclose( ApiLibHandle );
+            LibHandle    = NULL;
+            ApiLibHandle = NULL;
+            return FALSE;
+        }
     }
 
     return TRUE;
```

&nbsp;
_Thanks!_
___________
p.s. I could not get either `rexx -e "say .rexxinfo~version"` or `rexx -e "say .rexxinfo~majorversion"` to work for me, but then maybe that's because I'm still using 4.2.0? Is that something that's new with v5.0.0?

**Comment by Fish-Git** [comment]
Fixed by commit 271fb710dc3cf03dddbc1424859dba4fd13e4ff9.

Closing...

**Comment by Fish-Git** [comment]
_(Oops!)_

Bad commit!

Subsequent follow-on commit 7a9a89fc98df91b3fee7f5691b3bfd2de0f5fd2d fixes the missing statement in previous commit.
